### PR TITLE
Fix FUNC_ENTER macro mismatch in H5VL_object_inc_rc()

### PR DESCRIPTION
**Pull request by bazimo**
Full details in the issue.

### CODE REVIEW

**Review by derobins — CHANGES_REQUESTED** [review]
We should fix H5FD__supports_swmr_test() instead of making this change.

**Comment by bazimo** [comment]
> We should fix H5FD__supports_swmr_test() instead of making this change.

I attempted to change the macro in that function, but that broke another test. I wasn't sure which to use. I'm not really familiar with this code base; just ran in to these issues when running make check.

**Comment by derobins** [comment]
> > We should fix H5FD__supports_swmr_test() instead of making this change.
> 
> I attempted to change the macro in that function, but that broke another test. I wasn't sure which to use. I'm not really familiar with this code base; just ran in to these issues when running make check.

Let me take a look at it tonight.

File: src/H5VLint.c [hunk 1]
```diff
@@ -1028,7 +1028,7 @@ H5VL_object_inc_rc(H5VL_object_t *vol_obj)
     HDassert(vol_obj);
```
Comment (derobins) [inline src/H5VLint.c:1028]: Remove _NOFS from the FUNC_ENTER macro above, switch the FUNC_LEAVE macro back (NOFS was probably a copy/paste error), then revert the H5open() calls you added.

See if that works for you. Weird that we don't see failures on any of our systems(?)

Comment (bazimo) [reply src/H5VLint.c:1028]: Adjusting the macros solved that issue. Note that this PR is actually TWO problems; the second one that required H5open calls being added is not related to the H5VL_object_inc_rc function.

The other problem is that H5FD__supports_swmr_test does not initialize. However, I tried changing H5FD__supports_swmr_test (removed NOINIT from opening macro), and that didn't make a difference. The issue is the pthread storage key is still not initialized (if H5open call is removed):

[bappel@v4dev0]/bappel/hdf5-1.14.1_debug> r gdb
gdb test/.libs/lt-accum_swmr_reader
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-115.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /home/bappel/hdf5-1.14.1_debug/test/.libs/lt-accum_swmr_reader...done.
(gdb) b main
Breakpoint 1 at 0x40215d: file ../../hdf5/test/accum_swmr_reader.c, line 46.
(gdb) run
Starting program: /home/bappel/hdf5-1.14.1_debug/test/.libs/lt-accum_swmr_reader
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, main () at ../../hdf5/test/accum_swmr_reader.c:46
46          hid_t    fid  = -1;   /* File ID */
Missing separate debuginfos, use: debuginfo-install glibc-2.17-292.el7.x86_64 zlib-1.2.7-18.el7.x86_64
(gdb) n
47          hid_t    fapl = -1;   /* file access property list ID */
(gdb)
48          H5F_t   *f    = NULL; /* File pointer */
(gdb)
53          char    *driver         = NULL;  /* VFD string (from env variable) */
(gdb)
54          hbool_t  api_ctx_pushed = FALSE; /* Whether API context pushed */
(gdb)
59          driver = HDgetenv(HDF5_DRIVER);
(gdb)
61          if (!H5FD__supports_swmr_test(driver))
(gdb) s
H5FD__supports_swmr_test (vfd_name=0x0) at ../../hdf5/src/H5FDtest.c:94
94          hbool_t ret_value = FALSE;
(gdb) n
96          FUNC_ENTER_NOAPI_NOERR
(gdb)
98          if (!vfd_name || !HDstrcmp(vfd_name, "") || !HDstrcmp(vfd_name, "nomatch"))
(gdb)
99              ret_value = TRUE;
(gdb)
103         FUNC_LEAVE_NOAPI(ret_value)
(gdb)
lt-accum_swmr_reader: ../../hdf5/src/H5CS.c:223: H5CS_pop: Assertion `fstack->nused > 0' failed.

Program received signal SIGABRT, Aborted.
0x00007ffff660d377 in raise () from /lib64/libc.so.6
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/bappel/hdf5-1.14.1_debug/test/.libs/lt-accum_swmr_reader
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".

Breakpoint 1, main () at ../../hdf5/test/accum_swmr_reader.c:46
46          hid_t    fid  = -1;   /* File ID */
(gdb) n
47          hid_t    fapl = -1;   /* file access property list ID */
(gdb)
48          H5F_t   *f    = NULL; /* File pointer */
(gdb)
53          char    *driver         = NULL;  /* VFD string (from env variable) */
(gdb)
54          hbool_t  api_ctx_pushed = FALSE; /* Whether API context pushed */
(gdb)
59          driver = HDgetenv(HDF5_DRIVER);
(gdb)
61          if (!H5FD__supports_swmr_test(driver))
(gdb) s
H5FD__supports_swmr_test (vfd_name=0x0) at ../../hdf5/src/H5FDtest.c:94
94          hbool_t ret_value = FALSE;
(gdb) n
96          FUNC_ENTER_NOAPI_NOERR
(gdb) s
0x00007ffff66d7b40 in profil_counter () from /lib64/libc.so.6
(gdb)
Single stepping until exit from function profil_counter,
which has no line number information.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb)
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff740bcaf in H5CS_push (func_name=0x5 <Address 0x5 out of bounds>) at ../../hdf5/src/H5CS.c:170
170     {
(gdb) n

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66d8774 in mcount () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function mcount,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66d79f1 in __mcount_internal () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __mcount_internal,
which has no line number information.
0x00007ffff66d87a4 in mcount () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function mcount,
which has no line number information.
H5CS_push (func_name=0x7ffff7ac4110 <__func__.16567> "H5FD__supports_swmr_test") at ../../hdf5/src/H5CS.c:171
171         H5CS_t *fstack = H5CS_get_my_stack(); /* Current function stack for library */
(gdb) s

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff740b864 in H5CS__get_stack () at ../../hdf5/src/H5CS.c:81
81      {
(gdb) n

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66d8774 in mcount () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function mcount,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66d79f1 in __mcount_internal () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __mcount_internal,
which has no line number information.
0x00007ffff66d87a4 in mcount () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function mcount,
which has no line number information.
H5CS__get_stack () at ../../hdf5/src/H5CS.c:84
84          FUNC_ENTER_PACKAGE_NOERR_NOFS
(gdb) s

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b7 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff66066b0 in __ctype_b_loc () from /lib64/libc.so.6
(gdb) n
Single stepping until exit from function __ctype_b_loc,
which has no line number information.
H5CS__get_stack () at ../../hdf5/src/H5CS.c:86
86          fstack = H5TS_get_thread_local_value(H5TS_funcstk_key_g);
(gdb) p H5TS_funcstk_key_g
$1 = 0
(gdb)

Comment (derobins) [reply src/H5VLint.c:1028]: I'll switch the SWMR VFD check to initialize the library in a separate PR. Should be up soon and get merged today. After that, your these changes should work for you once you merge/rebase.

Comment (bazimo) [reply src/H5VLint.c:1028]: Acknowledged. I have removed the H5open calls from both PRs (1_14_1 & develop). Thank you for your assistance.

Comment (derobins) [reply src/H5VLint.c:1028]: I took a peek and changing other code shouldn't be necessary. I also can't reproduce your problem. Let's move the reproducibility conversation over to #2990.

File: test/accum_swmr_reader.c [hunk 2]
```diff
@@ -57,6 +57,7 @@ main(void)
      * by the environment variable.
      */
     driver = HDgetenv(HDF5_DRIVER);
+
```
Comment (derobins) [inline test/accum_swmr_reader.c]: Can you remove the blank line changes? Then we'll at least get this merged in while we discuss #2900.

Comment (bazimo) [reply test/accum_swmr_reader.c]: Done.

**Review by derobins — APPROVED** [review]

**Review by jhendersonHDF — APPROVED** [review]

**Pull request merged**